Sources/NIO/ChannelPipeline.swift

```diff
@@ -496,14 +496,17 @@ public final class ChannelPipeline: ChannelInvoker {
     func removeHandlers() {
         assert(eventLoop.inEventLoop)
 
-        if let head = self.head {
-            while let ctx = head.next {
-                remove0(ctx: ctx, promise: nil)
-            }
-            remove0(ctx: self.head!, promise: nil)
+        while let ctx = head.next {
+            remove0(ctx: ctx, promise: nil)
         }
-        self.head = nil
-        self.tail = nil
+
+        // After we removed all "user" errors we nil out (almost) all prev/next referenced to ensure we not have a reference cycle.
+        // We still set head.next = tail as this ensures we always can delegate to ChanneCore.* operations which may need to perform
+        // some special operations (like do some cleanup for received messages).
+        self.head.next = tail
```

Comment:
that looks good now to me. You might want to verify with Xcode's memory graph debugger